### PR TITLE
Staticize CASCADE_PROPERTIES, (temporarily) fix stylo path for animations, and introduce the long-term path to follow

### PR DESCRIPTION
**Pull request by emilio**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

- [x] These changes do not require tests because refactoring + geckolib

@bholley: I was going to do this "the good way", but that involves quite a few properties, so I thought I'd rather unlock stylo before :)

r? @bholley

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11972)

### CODE REVIEW

**Comment by bholley** [comment]
I don't understand why we introduce need_borrow and then remove it again?

**Comment by bholley** [comment]
(the reason I did clone instead of borrow is that borrowing will almost never be possible from geckolib, because we need to convert the types)

**Comment by emilio** [comment]
Yup, I'm doing some changes, sorry for the early review request

**Comment by emilio** [comment]
Ok, now it should be ready for review.

r? @bholley

File: components/style/properties/longhand/box.mako.rs [hunk 1]
```diff
@@ -343,7 +344,9 @@ ${helpers.single_keyword("overflow-x", "visible hidden scroll auto",
 
 // TODO(pcwalton): Lots more timing functions.
 // TODO(pcwalton): Multiple transitions.
-<%helpers:longhand name="transition-timing-function" animatable="False">
+<%helpers:longhand name="transition-timing-function"
+                   need_index="True"
+                   animatable="False">
```
Comment (bholley) [inline components/style/properties/longhand/box.mako.rs]: Nit - the order here isn't consistent.

File: ports/geckolib/properties.mako.rs [hunk 2]
```diff
@@ -265,14 +293,20 @@ def set_gecko_property(ffi_name, expr):
...
+<%def name="impl_simple(ident, gecko_ffi_name, need_clone=False)">
 <%call expr="impl_simple_setter(ident, gecko_ffi_name)"></%call>
 <%call expr="impl_simple_copy(ident, gecko_ffi_name)"></%call>
+% if need_clone:
+    <%call expr="impl_simple_clone(ident, gecko_ffi_name)"></%call>
+% endif
 </%def>
 
-<%def name="impl_color(ident, gecko_ffi_name, color_flags_ffi_name=None)">
+<%def name="impl_color(ident, gecko_ffi_name, color_flags_ffi_name=None, need_clone=None)">
```
Comment (bholley) [inline ports/geckolib/properties.mako.rs]: This should be False, not None.

File: ports/geckolib/properties.mako.rs [hunk 3]
```diff
@@ -245,15 +246,42 @@ def set_gecko_property(ffi_name, expr):
     }
 </%def>
 
+<%def name="impl_color_setter(ident, gecko_ffi_name, color_flags_ffi_name=None)">
```
Comment (bholley) [inline ports/geckolib/properties.mako.rs]: Looks like this is a duplicate of the above - remove it?

File: ports/geckolib/properties.mako.rs [hunk 4]
```diff
@@ -245,15 +246,42 @@ def set_gecko_property(ffi_name, expr):
...
+            Color::RGBA(rgba) => convert_rgba_to_nscolor(&rgba),
+        };
+        ${set_gecko_property(gecko_ffi_name, "result")}
+    }
+</%def>
+
 <%def name="impl_color_copy(ident, gecko_ffi_name, color_flags_ffi_name=None)">
     fn copy_${ident}_from(&mut self, other: &Self) {
         % if color_flags_ffi_name:
-        ${clear_color_flags(color_flags_ffi_name)}
```
Comment (bholley) [inline ports/geckolib/properties.mako.rs]: Why are we removing the call to clear_color_flags?

File: ports/geckolib/values.rs [hunk 5]
```diff
@@ -12,65 +12,147 @@ use style::values::computed::{LengthOrPercentage, LengthOrPercentageOrAuto, Leng
...
+    }
+    #[inline]
+    fn get_factor(&self) -> f32 {
+        debug_assert!(self.is_factor());
+        unsafe { *self.mValue.mFloat.as_ref() }
+    }
 }
 
-pub trait ToGeckoStyleCoord {
+pub trait ToGeckoStyleCoord : Sized {
```
Comment (bholley) [inline ports/geckolib/values.rs]: Might want to rename this to GeckoStyleCoordConversion, or some potentially-more-idiomatic name that reflects the bidirectional conversion.

File: ports/geckolib/properties.mako.rs (hunk 4 above)
Comment (emilio) [inline ports/geckolib/properties.mako.rs]: Whoops, good catch. At one point I wrote another version of these functions, but then backed it out, and this seems that it got lost in the meantime.

File: ports/geckolib/properties.mako.rs (hunk 3 above)
Comment (emilio) [inline ports/geckolib/properties.mako.rs]: Same reason, my fault

File: ports/geckolib/properties.mako.rs (hunk 2 above)
Comment (emilio) [inline ports/geckolib/properties.mako.rs]: Ack'd

**Comment by bholley** [comment]
r=me with those fixes.

@bors-servo delegate+

**Comment by emilio** [comment]
These were legit failures (`get_mod` did `count % element` instead of the opposite). It's fixed.

@bors-servo: r=bholley

**Comment by emilio** [comment]
@bors-servo: retry

 * https://github.com/servo/servo/issues/11574

**Comment by emilio** [comment]
@bors-servo: retry

* iframe-parentage (I can't look up the issue number right now)

**Comment by ConnorGBrewster** [comment]
@bors-servo retry
 * timeout?

**Comment by ConnorGBrewster** [comment]
@bors-servo retry
 * #11574

**Comment by ConnorGBrewster** [comment]
@bors-servo retry
 * #11703

**Comment by jdm** [comment]
@bors-servo: retry
* #11703

**Pull request merged**